src/java/org/apache/cassandra/service/MigrationCoordinator.java

```diff
@@ -0,0 +1,512 @@
+    private static final RuntimeMXBean runtimeMXBean = ManagementFactory.getRuntimeMXBean();
+    private static final Future<Void> FINISHED_FUTURE = Futures.immediateFuture(null);
+
+    private static final int MIGRATION_DELAY_IN_MS = 60000;
+    private static final int MAX_OUTSTANDING_VERSION_REQUESTS = 3;
+
+    public static final MigrationCoordinator instance = new MigrationCoordinator();
+
+    static class VersionInfo
+    {
+        final UUID version;
+
+        final Set<InetAddress> endpoints           = Sets.newConcurrentHashSet();
+        final Set<InetAddress> outstandingRequests = Sets.newConcurrentHashSet();
+        final Deque<InetAddress> requestQueue      = new ArrayDeque<>();
+
+        private final WaitQueue waitQueue = new WaitQueue();
+
+        volatile boolean receivedSchema;
+
+        VersionInfo(UUID version)
+        {
+            this.version = version;
+        }
+
+        WaitQueue.Signal register()
+        {
+            return waitQueue.register();
+        }
+
+        void markReceived()
+        {
+            if (receivedSchema)
+                return;
+
+            receivedSchema = true;
+            waitQueue.signalAll();
+        }
+
+        boolean wasReceived()
+        {
+            return receivedSchema;
+        }
+    }
+
+    private final Map<UUID, VersionInfo> versionInfo = new HashMap<>();
+    private final Map<InetAddress, UUID> endpointVersions = new HashMap<>();
+
+    public void start()
+    {
+        ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(this::pullUnreceivedSchemaVersions, 1, 1, TimeUnit.MINUTES);
+    }
+
+    public synchronized void reset()
+    {
+        versionInfo.clear();
+    }
+
+    synchronized List<Future<Void>> pullUnreceivedSchemaVersions()
+    {
+        List<Future<Void>> futures = new ArrayList<>();
+        for (VersionInfo info : versionInfo.values())
+        {
+            if (info.wasReceived() || info.outstandingRequests.size() > 0)
+                continue;
+
+            Future<Void> future = maybePullSchema(info);
+            if (future != null && future != FINISHED_FUTURE)
+                futures.add(future);
+        }
+
+        return futures;
+    }
+
+    synchronized Future<Void> maybePullSchema(VersionInfo info)
+    {
+        if (info.endpoints.isEmpty() || info.wasReceived() || !shouldPullSchema(info.version))
+            return FINISHED_FUTURE;
+
+        if (info.outstandingRequests.size() >= getMaxOutstandingVersionRequests())
+            return FINISHED_FUTURE;
+
+        for (int i=0, isize=info.requestQueue.size(); i<isize; i++)
+        {
+            InetAddress endpoint = info.requestQueue.remove();
+            if (!info.endpoints.contains(endpoint))
+                continue;
+
+            if (shouldPullFromEndpoint(endpoint) && info.outstandingRequests.add(endpoint))
+            {
+                return scheduleSchemaPull(endpoint, info);
+            }
```

Comment:
here, `scheduleSchemaPull` is only called if the endpoint is not already in `outstandinRequests`, so I believe this effectively does the same thing as #95